File: backend/controller/user.controller.js
import { User } from "../models/user.model.js";
import bcrypt from "bcryptjs";
import jwt from "jsonwebtoken";

export const register = async (req, res) => {
    try {
        const { fullName, username, email, password, phoneNumber, role } = req.body;
        
        if (!fullName || !username || !email || !password || !phoneNumber || !role) {
            return res.status(400).json({
                message: "All fields are required",
                success: false
            });
        }

        // Validate role
        if (role !== 'customer' && role !== 'agent') {
            return res.status(400).json({
                message: "Role must be either 'customer' or 'agent'",
                success: false
            });
        }

        // Check if user already exists
        const existingUser = await User.findOne({ $or: [{ email }, { username }] });
        if (existingUser) {
            return res.status(400).json({
                message: "User already exists with this email or username",
                success: false
            });
        }

        // Hash password
        const hashedPassword = await bcrypt.hash(password, 10);

        // Create new user
        await User.create({
            fullName,
            username,
            email,
            password: hashedPassword,
            phoneNumber,
            role
        });

        return res.status(201).json({
            message: "Account created successfully",
            success: true
        });

    } catch (error) {
        console.log(error);
        return res.status(500).json({
            message: "Server error",
            success: false
        });
    }
};

export const login = async (req, res) => {
    try {
        const { email, username, password } = req.body;

        if ((!email && !username) || !password) {
            return res.status(400).json({
                message: "Username/email and password are required",
                success: false
            });
        }

        // Find user by email or username
        let user = await User.findOne({ $or: [{ email }, { username }] });
        if (!user) {
            return res.status(400).json({
                message: "Incorrect username/email or password",
                success: false
            });
        }

        // Check if password matches
        const isPasswordMatch = await bcrypt.compare(password, user.password);
        if (!isPasswordMatch) {
            return res.status(400).json({
                message: "Incorrect username/email or password",
                success: false
            });
        }

        // Generate JWT token
        const tokenData = {
            userId: user._id
        };
        const token = jwt.sign(tokenData, process.env.SECRET_KEY, { expiresIn: '1d' });

        // Send response with cookie
        user = {
            _id: user._id,
            fullName: user.fullName,
            username: user.username,
            email: user.email,
            phoneNumber: user.phoneNumber,
            role: user.role
        };

        return res.status(200).cookie("token", token, { 
            maxAge: 1 * 24 * 60 * 60 * 1000, // 1 day
            httpOnly: true, 
            sameSite: 'strict' 
        }).json({
            message: `Welcome back ${user.fullName}`,
            user,
            success: true
        });

    } catch (error) {
        console.log(error);
        return res.status(500).json({
            message: "Server error",
            success: false
        });
    }
};

export const logout = async (req, res) => {
    try {
        return res.status(200).cookie("token", "", { maxAge: 0 }).json({
            message: "Logged out successfully",
            success: true
        });
    } catch (error) {
        console.log(error);
        return res.status(500).json({
            message: "Server error",
            success: false
        });
    }
};

File: backend/controllers/user_controller.py
from flask import request, jsonify, make_response
from models.user_model import user_model
from bson.objectid import ObjectId
import bcrypt
import jwt
import os
import logging
from datetime import datetime, timedelta

logger = logging.getLogger(__name__)

def register():
    """Register a new user"""
    try:
        data = request.get_json()
        
        full_name = data.get('fullName')
        username = data.get('username')
        email = data.get('email')
        password = data.get('password')
        phone_number = data.get('phoneNumber')
        role = data.get('role')
        
        # Validate all fields
        if not all([full_name, username, email, password, phone_number, role]):
            return jsonify({
                'message': 'All fields are required',
                'success': False
            }), 400
        
        # Validate role
        if role not in ['customer', 'agent']:
            return jsonify({
                'message': "Role must be either 'customer' or 'agent'",
                'success': False
            }), 400
        
        # Check if user already exists
        existing_user = user_model.find_existing(email, username)
        if existing_user:
            return jsonify({
                'message': 'User already exists with this email or username',
                'success': False
            }), 400
        
        # Hash password using optimized salt rounds
        hashed_password = bcrypt.hashpw(password.encode('utf-8'), bcrypt.gensalt(12))
        
        # Create new user
        user_model.create({
            'fullName': full_name,
            'username': username,
            'email': email,
            'password': hashed_password,
            'phoneNumber': phone_number,
            'role': role
        })
        
        logger.info(f"User registered: {email} (role: {role})")
        
        return jsonify({
            'message': 'Account created successfully',
            'success': True
        }), 201
    
    except Exception as error:
        logger.error(f"Registration error: {error}", exc_info=True)
        return jsonify({
            'message': 'Server error',
            'success': False
        }), 500

def login():
    """Login user"""
    try:
        data = request.get_json()
        
        email = data.get('email')
        username = data.get('username')
        password = data.get('password')
        
        # Validate credentials
        if (not email and not username) or not password:
            return jsonify({
                'message': 'Username/email and password are required',
                'success': False
            }), 400
        
        # Find user
        user = user_model.find_one({'$or': [{'email': email}, {'username': username}]})
        
        if not user:
            logger.warning(f"Login attempt failed: user not found ({email or username})")
            return jsonify({
                'message': 'Incorrect username/email or password',
                'success': False
            }), 400
        
        # Check password
        if not bcrypt.checkpw(password.encode('utf-8'), user['password']):
            logger.warning(f"Login attempt failed: incorrect password ({email or username})")
            return jsonify({
                'message': 'Incorrect username/email or password',
                'success': False
            }), 400
        
        # Generate JWT token
        token_data = {
            'userId': str(user['_id']),
            'exp': datetime.utcnow() + timedelta(hours=24)
        }
        token = jwt.encode(
            token_data,
            os.getenv('SECRET_KEY'),
            algorithm='HS256'
        )
        
        # Prepare response
        response_data = {
            'message': f"Welcome back {user['fullName']}",
            'user': {
                '_id': str(user['_id']),
                'fullName': user['fullName'],
                'username': user['username'],
                'email': user['email'],
                'phoneNumber': user['phoneNumber'],
                'role': user['role']
            },
            'success': True
        }
        
        logger.info(f"User logged in: {email or username}")
        
        response = make_response(jsonify(response_data), 200)
        
        # Set cookie with secure flag only in production
        is_production = os.getenv('FLASK_ENV') == 'production'
        
        response.set_cookie(
            'token',
            token,
            max_age=24 * 60 * 60,
            httponly=True,
            secure=is_production,  # ? HTTPS only in production
            samesite='None' if is_production else 'Lax'  # ? More lenient in dev
        )
        
        return response
    
    except Exception as error:
        logger.error(f"Login error: {error}", exc_info=True)
        return jsonify({
            'message': 'Server error',
            'success': False
        }), 500

def logout():
    """Logout user"""
    try:
        response = make_response(jsonify({
            'message': 'Logged out successfully',
            'success': True
        }), 200)
        
        response.set_cookie('token', '', max_age=0)
        logger.info("User logged out")
        
        return response
    
    except Exception as error:
        logger.error(f"Logout error: {error}", exc_info=True)
        return jsonify({
            'message': 'Server error',
            'success': False
        }), 500